driver/src/price_estimation/kraken.rs

```diff
@@ -1,4 +1,200 @@
 //! Implementation of a price source for Kraken.
 
-#[allow(dead_code)]
 mod api;
+
+use self::api::{Asset, AssetPair, KrakenApi, KrakenHttpApi};
+use super::{PriceSource, Token, TokenId};
+use anyhow::{anyhow, Context, Result};
+use std::collections::HashMap;
+
+/// A client to the Kraken exchange.
+pub struct KrakenClient<Api> {
+    /// A Kraken API implementation. This allows for mocked Kraken APIs to be
+    /// used for testing.
+    api: Api,
+}
+
+impl KrakenClient<KrakenHttpApi> {
+    /// Creates a new client instance using an HTTP API instance and the default
+    /// Kraken API base URL.
+    pub fn new() -> Result<Self> {
+        let api = KrakenHttpApi::new()?;
+        Ok(KrakenClient::with_api(api))
+    }
+}
+
+impl<Api> KrakenClient<Api>
+where
+    Api: KrakenApi,
+{
+    /// Create a new client instance from an API.
+    pub fn with_api(api: Api) -> Self {
+        KrakenClient { api }
+    }
+
+    /// Generates a mapping between Kraken asset pair identifiers and tokens
+    /// that are used when computing the price map.
+    fn get_token_asset_pairs<'a>(&self, tokens: &'a [Token]) -> Result<HashMap<String, &'a Token>> {
+        // TODO(nlordell): If these calls start taking too long, we can consider
+        //   caching this information somehow. The only thing that is
+        //   complicated is determining when the cache needs to be invalidated
+        //   as new assets get added to Kraken.
```

Comment:
Is there an open issue for this TODO?

Reply:
There was, but there was also a typo in the title: #583

Reply:
Not for this particular `TODO`. Maybe a `NOTE` would be better. I just left it there as a possible optimization if delays become an issue in the future.

There is a related case here: #583